### PR TITLE
Update breath detection algorithm.

### PR DESCRIPTION
**Pull request by jlebar**
<git-pr-chain>

#### Commits in this PR
1. Update breath detection algorithm.
    
    Instead of using instantaneous flows, use averages of flows.
    
    See details in code.
    
    Video in action: https://photos.app.goo.gl/p2WN3dGDYaB6h1TE6

#### [PR chain](https://github.com/jlebar/git-pr-chain)
1. 👉 #548 Update breath detection algorithm. 👈 **YOU ARE HERE**
1. #527 Open exhale pinch valve to 0.3 during PIP.
1. #528 Add fan pressure readings at various power levels.
1. #531 DRAFT - Use different fan speed depending on PIP.

</git-pr-chain>

### CODE REVIEW

**Comment by jlebar** [comment]
cc @inceptionev @sglow

File: controller/test/blower_fsm/blower_fsm_test_data.h [hunk 1]
```diff
@@ -0,0 +1,1061 @@
+/* Copyright 2020f,
+RespiraWorks
+Licensed under the Apache Licensef,
```
Comment (jkff) [inline controller/test/blower_fsm/blower_fsm_test_data.h:3]: Something's funky with this - why is there an extra "f" at the end of many words? (2020, License, writing, BASIS, KIND)

File: controller/test/blower_fsm/blower_fsm_test_data.h [hunk 2]
```diff
@@ -0,0 +1,1061 @@
...
+//  - PIP  20cm
+//  - PEEP 10cm
+//  - rr   12bpm (5s/breath)
+//  - i:e  0.25 (itime=1s, etime=4s)
+//
+// Duration between each line is 10ms.
+//
+// There's a manual inspiratory attempt in here; in other words, I expanded the
+// test lung a bit during the expiratory dwell.
+inline const VolumetricFlow FLOW_TRACE_1[] = {
```
Comment (jkff) [inline controller/test/blower_fsm/blower_fsm_test_data.h:33]: Will this data ever need to be updated? If yes, how would one do it? Or basically, how did you capture the data and create this .h file?

File: controller/test/blower_fsm/blower_fsm_test_data.h [hunk 3]
```diff
@@ -0,0 +1,1061 @@
...
+// preset #2:
+//
+//  - PIP  20cm
+//  - PEEP 10cm
+//  - rr   12bpm (5s/breath)
+//  - i:e  0.25 (itime=1s, etime=4s)
+//
+// Duration between each line is 10ms.
+//
+// No manual inspiratory effort here.
```
Comment (jkff) [inline controller/test/blower_fsm/blower_fsm_test_data.h:553]: Reflect this in the name of the variable?

File: controller/lib/core/blower_fsm.h [hunk 4]
```diff
@@ -178,11 +172,18 @@ class PressureAssistFsm {
   Time inspire_end_;
   Time expire_deadline_;
 
-  // initialize threshold at insanely high flow to not trigger breath before
-  // a proper threshold has been computed
-  VolumetricFlow inspiratory_effort_threshold_ = cubic_m_per_sec(9);
   bool inspire_finished_ = false;
   bool finished_ = false;
+
+  // During exhale we maintain two exponentially-weighted averages of flow, one
```
Comment (jkff) [inline controller/lib/core/blower_fsm.h:178]: Please link the Slack thread where this algorithm was discussed.

File: controller/lib/core/blower_fsm.cpp [hunk 5]
```diff
@@ -15,8 +15,40 @@ limitations under the License.
...
+
+// TODO: This should be configurable from the GUI.
+static DebugFloat dbg_pa_flow_trigger("pa_flow_trigger",
+                                      "pressure assist flow trigger (ml/s)",
+                                      200);
+
+// TODO: Is 250ms right?  Or can it be a fixed value at all; should it depend
+// on the RR or something?
+static DebugFloat
+    dbg_pa_min_expire_ms("pa_min_dwell_ms",
```
Comment (jkff) [inline controller/lib/core/blower_fsm.cpp:33]: should it be called dbg_pa_min_dwell_ms?

File: controller/lib/core/blower_fsm.cpp [hunk 6]
```diff
@@ -15,8 +15,40 @@ limitations under the License.
...
+                         "minimum amount of time after ventilator exits PIP "
+                         "before we're eligible to trigger a breath",
+                         250);
+
+// pa_fast_avg_alpha and pa_slow_avg_alpha were tuned for a control loop that
+// runs at a particular frequency.  If the control loop gets slower, the alpha
```
Comment (jkff) [inline controller/lib/core/blower_fsm.cpp:39]: Code below seems to suggest that these variables actually will not need tuning if the loop period changes

File: controller/lib/core/blower_fsm.cpp [hunk 7]
```diff
@@ -105,34 +137,40 @@ BlowerSystemState PressureAssistFsm::DesiredState() const {
...
+  // one.  You can think of the slow average as estimating "flow at dwell" and
+  // the fast average as estimating "current flow".
+  //
+  // If the fast average exceeds the slow average, we trigger a breath.
+  float slow_alpha = dbg_pa_slow_avg_alpha.Get();
+  float fast_alpha = dbg_pa_fast_avg_alpha.Get();
+
+  // TODO: This could be encapsulated in an exponentially-weighted-average
+  // class.
+  if (slow_avg_flow_ == std::nullopt) {
```
Comment (jkff) [inline controller/lib/core/blower_fsm.cpp:160]: You can replace these with .value_or() at lines 167 and 169.

File: controller/test/blower_fsm/blower_fsm_test.cpp [hunk 8]
```diff
@@ -62,17 +66,35 @@ void testSequence(
                    /*expected_pressure_setpoint*/ std::optional<Pressure>,
                    /*expected_flow_direction*/ FlowDirection>> &seq) {
   BlowerFsm fsm;
+  if (seq.empty()) {
+    return;
+  }
+
+  // Reset time to test's start time.
+  Hal.delay(microseconds(-Hal.now().microsSinceStartup()));
```
Comment (jkff) [inline controller/test/blower_fsm/blower_fsm_test.cpp:74]: Uh a negative delay looks super awkward. Can we do without it? (OK to leave a TODO explaining how)

File: controller/test/blower_fsm/blower_fsm_test.cpp [hunk 9]
```diff
@@ -88,142 +110,157 @@ TEST(BlowerFsmTest, PressureControl) {
...
-       FlowDirection::INSPIRATORY},
-      {p, inputs_breath, 5200 + rise_time_ms, cmH2O(20),
-       FlowDirection::INSPIRATORY},
-      // triggered breath is of normal duration
-      {p, inputs_zero, 7199, cmH2O(20), FlowDirection::INSPIRATORY},
-      {p, inputs_zero, 7201, cmH2O(10), FlowDirection::EXPIRATORY},
-  });
+  ASSERT_TRUE(results.expire_start_time.has_value());
+  ASSERT_TRUE(results.finish_time.has_value());
+  EXPECT_EQ(results.expire_start_time->milliseconds(), 1000.f);
```
Comment (jkff) [inline controller/test/blower_fsm/blower_fsm_test.cpp:228]: Expecting exact equality seems fishy, is this dependent on the algorithm or only on the setpoint checks?

File: controller/test/blower_fsm/blower_fsm_test_data.h (hunk 1 above)
Comment (jlebar) [reply controller/test/blower_fsm/blower_fsm_test_data.h:3]: Oops, this was my find-and-replace gone very awry.  Thanks.

File: controller/test/blower_fsm/blower_fsm_test_data.h (hunk 2 above)
Comment (jlebar) [reply controller/test/blower_fsm/blower_fsm_test_data.h:33]: Expanded the comment a bit, although I'm not sure if this is enough.  LMK if there's anything else missing and I can go back and add it

```
// Trace of net flow readings captured on a live device using pressure command
// mode at covent preset #2:
//
//  - PIP  20cm
//  - PEEP 10cm
//  - rr   12bpm (5s/breath)
//  - i:e  0.25 (itime=1s, etime=4s)
//
// Duration between each line is 10ms.
//
// There's a manual inspiratory attempt in here; in other words, I expanded the
// test lung a bit with my finger during the expiratory dwell.
```

File: controller/lib/core/blower_fsm.cpp (hunk 7 above)
Comment (jlebar) [reply controller/lib/core/blower_fsm.cpp:160]: Much better, thanks!

File: controller/test/blower_fsm/blower_fsm_test.cpp (hunk 9 above)
Comment (jlebar) [reply controller/test/blower_fsm/blower_fsm_test.cpp:228]: In pressure control the machine controls the duration of inspiration and expiration precisely.  In pressure assist, the patient controls the expire duration, but the machine still controls the inspire duration.

So I think this is basically correct -- if I understand what you're questioning.

**Comment by jlebar** [comment]
Thank you for the review, and sorry this was a bit of a mess.

**Pull request merged**